src/gluonts/core/context.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+gluonts.core.context
+~~~~~~~~~~~~~~~~~~~~
+
+This modules offers a `Context`-class, which allows to manage a global context.
+
+The idea is to support a form of dependency injection, where instead of passing
+a concrete value along the call-chain, it is shared through the context.
+
+`gluonts.env` is such a context and is used to manage global settings, such as
+the number of workers in multiprocessing.
+
+Example::
+
+    from gluonts.core.context import Context
+
+    class MyContext(Context):
+        debug: bool = False
+
+    def fn():
+        if ctx.debug:
+            print("In debug mode.")
+
+    # we use `_let` instead of `let` to avoid possible name-collisions
+    # so you can do `ctx._let(let=...)`
+    with ctx._let(debug=True):
+        # this will print the message
+        fn()
+
+    # no message will be printed
+    fn()
+
+Another option is to inject the context to a function. This has the advantage,
+that you can still manually pass values, but use the context as a fallback::
+
+    @ctx._inject("debug")
+    def fn(debug):
+        ...
+
+    # this will use the value defined in the context
+    fn()
+
+    # but one can still set the value manually
+    fn(False)
+"""
+
+import functools
+import inspect
+from typing import Any
+
+import pydantic
+
+
+class _Config:
+    arbitrary_types_allowed = True
+
+
+class Context:
```

Comment:
I'm wondering if this could create confusion (there's python context managers, mxnet's computational context...)

Would `Namespace` be a better name?

Reply:
I see the possible confusion, but I think `Context` is still a better name than `Namespace`. We could name it `Setting` though.